File: Software Engineering Foundations/tic_tac_toe_v2/game.rb
require './board.rb'
require './human_player.rb'

class Game
    def initialize(size, *marks)
        @players = marks.map{|mark| HumanPlayer.new(mark)}
        @current_player = @players.first
        @board = Board.new(size)
    end

    def switch_turn
        @players.rotate!(1)
        @current_player = @players.first
    end

    def play
        while @board.empty_positions?
            @board.print
            pos = @current_player.get_position
            @board.place_mark(pos, @current_player.mark)
            if @board.win?(@current_player.mark)
                puts 'Game Over'
                puts "Player #{@current_player.mark} has won!"
                return
            else
                self.switch_turn
            end
        end
        puts "Game Over"
        puts "Draw"
    end
end

File: Javascript/1_Event_Loop/tic_tac_toe/game.js
const Board = require('./board.js');

class Game {
    constructor (level) {
        this.board = new Board (level);
        this.currentPlayer = Board.marks[0]
    }

    isOver () {
        return this.board.isOver();
    }

    promptMove (reader, callback) {
        this.board.print()
        console.log(`Current Player > ${this.currentPlayer}`)

        reader.question('Enter row > ', (row) => {
            const x = parseInt(row);
            reader.question('Enter col > ', (col) => {
                const y = parseInt(col);
                const pos = [x, y];
                callback(pos);
            })
        })
    }

    playMove (pos) {
        const move = this.board.place_mark(pos, this.currentPlayer);
        if (move) {
            this.swapTurn();
            return true;
        } else {
            return false;
        }
    }

    swapTurn () {
        this.currentPlayer = this.currentPlayer === Board.marks[0] ? Board.marks[1] : Board.marks[0];
        return this.currentPlayer
    }

    winner () {
        return this.board.winner;
    }

    run (reader, reset) {
        // make a move
        this.promptMove (reader, pos => {
            this.playMove(pos);

            // if over
            if (this.isOver()) {
                this.board.print();
                if (this.winner()) {
                    console.log(`Player ${this.winner()} has won!`);
                } else {
                    console.log('It\'s a tie!');
                }
                reset()
            } else {
                this.run(reader, reset);
            }
        })
    }
}

module.exports = Game;